tsdb/db.go

```diff
@@ -597,6 +597,92 @@ func (db *DBReadOnly) Blocks() ([]BlockReader, error) {
 	return blockReaders, nil
 }
 
+// LastBlockID returns the BlockID of latest block.
+func (db *DBReadOnly) LastBlockID(logger log.Logger) (string, error) {
+	select {
+	case <-db.closed:
+		return "", ErrClosed
+	default:
+	}
+	latestDirName, err := db.lastBlockDirName(logger)
+	if err != nil {
+		return "", err
+	}
+	// Open the latest block and get its ID
+	block, err := db.Block(logger, latestDirName)
+	if err != nil {
+		return "", err
+	}
+	return block.Meta().ULID.String(), nil
+}
+
+func (db *DBReadOnly) lastBlockDirName(logger log.Logger) (string, error) {
+	latestDirName := ""
+
+	// Walk the blocks directory and find the latest subdirectory
+	err := filepath.Walk(filepath.Join(db.dir), func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if !info.IsDir() {
+			return nil
+		}
+		dirName := filepath.Base(path)
+
+		// Parse the directory name as a timestamp
+		entries, err := os.ReadDir(dirName)
+		if err != nil {
+			return nil
+		}
+		max := uint64(0)
+		for _, e := range entries {
+			if e.Name() == "index" || e.Name() == "chunks_head" || e.Name() == "lock" || e.Name() == "queries.active" || e.Name() == "wal" {
```

Comment:
Instead of calling e.name() multiple times here as well as down inside the loop (#lines 639, 644, 652), we can call it once and a `set` of all the elements to check against